pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -17,74 +17,148 @@ limitations under the License.
 package aggregation
 
 import (
+	"math"
 	"sync"
 	"time"
+
+	"github.com/davecgh/go-spew/spew"
 )
 
 // TimedFloat64Buckets keeps buckets that have been collected at a certain time.
 type TimedFloat64Buckets struct {
 	bucketsMutex sync.RWMutex
-	// Metrics received in a certain timeframe are all summed up.
-	// This assumes that we don't take multiple readings of
-	// the same metric in the same bucket (per second currently).
-	// The only case where this might happen currently is when activator scales
-	// a revision from 0. The metrics for that bucket might be off
-	// by exactly "1" as that poke always reports a concurrency of 1.
-	// Since we're windowing metrics anyway, that slight skew is acceptable.
-	buckets map[time.Time]float64
+	buckets      []float64
+	lastWrite    time.Time
 
 	granularity time.Duration
+	window      time.Duration
+}
+
+// Implements stringer interface.
+func (t *TimedFloat64Buckets) String() string {
+	return spew.Sdump(t.buckets)
 }
 
-// NewTimedFloat64Buckets generates a new TimedFloat64Buckets with the given
+// NewTimedFloat64Buckets generates a new TimedFloat64Buckets2 with the given
 // granularity.
-func NewTimedFloat64Buckets(granularity time.Duration) *TimedFloat64Buckets {
+func NewTimedFloat64Buckets(window, granularity time.Duration) *TimedFloat64Buckets {
+	// Number of buckets is `window` divided by `granularity`, rounded up.
+	// e.g. 60s / 2s = 30.
+	nb := int(math.Ceil(float64(window) / float64(granularity)))
 	return &TimedFloat64Buckets{
-		buckets:     make(map[time.Time]float64),
+		buckets:     make([]float64, nb),
 		granularity: granularity,
+		window:      window,
+	}
+}
+
+// timeToIndex converts time to an integer that can be used for modulo
+// operations to find the index in the bucket list.
+// bucketMutex needs to be held.
+func (t *TimedFloat64Buckets) timeToIndex(tm time.Time) int {
+	// I don't think this run in 2038 :-)
+	// NB: we need to divide by granularity, since it's a compressing mapping
+	// to buckets.
+	return int(tm.Unix()) / int(t.granularity.Seconds())
+}
+
+func (t *TimedFloat64Buckets) reset() {
```

Comment:
This needs a comment stating the mutex needs to be held. I'd actually just inline this at its only occurrence 🤷‍♂